File: src/lib/economy/simulation.ts
import { BASELINE_POLICY, POLICIES, currentPolicy, solveModel, type ModelInputs, type ModelMode, type Objective, type Policy, type ProfileOutcome } from './pirates-model';
import { solvePackageElection, type ElectionSearch, type PackageElectionResult } from './package-election';
import type { PackageBallotResult } from './package-ballot';
export interface ScenarioRequest { id: number; inputs: ModelInputs; mode: ModelMode; foreignObjective: Objective; /** Exclude a pause from both actors’ policy menus. */ pauseUnavailable?: boolean; /** Legacy callers may supply this; deployment is now on the ballot. */ pace?: number }
export interface ScenarioSnapshot {
  inputs: ModelInputs; mode: ModelMode; foreignObjective: Objective; pauseUnavailable: boolean;
  selected: ProfileOutcome; statusQuo: ProfileOutcome; baseline: ProfileOutcome;
  /** At most eight highest-support funded packages, including the selected policy when it ranks. */
  alternatives: ProfileOutcome[]; leading?: ProfileOutcome;
  ballot: PackageBallotResult; selection: PackageElectionResult['selection'];
  policyCount: number; foreignPolicyCount: number; evaluations: number; search: ElectionSearch;
  foreignBestPolicy?: Policy; foreignBestResponseGain?: number;
}
export type ScenarioResponse = { id: number; snapshot: ScenarioSnapshot; error?: never } | { id: number; error: string; snapshot?: never };
/** Only selected, reference, and leading profiles cross the worker boundary. */
export function solveScenario(request: ScenarioRequest): ScenarioSnapshot {
  const model = solveModel(request.inputs, { mode: request.mode, objective: 'workers', foreignObjective: request.foreignObjective, pace: 1 });
  const pauseUnavailable = request.pauseUnavailable ?? false;
  const policies = pauseUnavailable ? POLICIES.filter(policy => policy.pace !== 0) : POLICIES;
  const foreignPolicies = pauseUnavailable ? model.foreignPolicies.filter(policy => policy.pace !== 0) : model.foreignPolicies;
  const election = solvePackageElection({ ...model, policies, foreignPolicies, currentPolicy: currentPolicy(1) });
  const cache = new Map<string, ProfileOutcome>();
  const materialize = (policy: Policy, foreign = election.foreignPolicy): ProfileOutcome => {
    const key = policy.id + '::' + (foreign?.id ?? 'none');
    let profile = cache.get(key);
    if (!profile) { profile = model.evaluate(policy, foreign); cache.set(key, profile); }
    return profile;
  };
  const byId = new Map(policies.map(policy => [policy.id, policy]));
  const ranked = [...election.ballot.tallies].sort((a, b) => b.supportPercent - a.supportPercent || (a.policyId < b.policyId ? -1 : a.policyId > b.policyId ? 1 : 0));
  const selected = materialize(election.usPolicy);
  const alternatives = ranked.slice(0, 8).map(tally => materialize(byId.get(tally.policyId)!));
  return {
    inputs: model.inputs, mode: model.mode, foreignObjective: model.foreignObjective, pauseUnavailable, selected,
    statusQuo: materialize(currentPolicy(1)),
    baseline: model.evaluate(BASELINE_POLICY, model.mode === 'strategic' ? BASELINE_POLICY : undefined),
    alternatives, leading: alternatives[0], ballot: election.ballot, selection: election.selection,
    policyCount: policies.length, foreignPolicyCount: foreignPolicies.length, evaluations: election.evaluations,
    search: election.search, foreignBestPolicy: election.foreignBestPolicy, foreignBestResponseGain: election.foreignBestResponseGain,
  };
}

File: calculator/simulation.py
"""Assemble a compact, JSON-safe scenario from the economic and voting models."""

from __future__ import annotations

from typing import Any

from .election import solve_package_election
from .model import solve_model
from .policies import BASELINE_POLICY, POLICIES, current_policy


def scenario_settings(request: dict[str, Any]) -> tuple[str, str, bool]:
    """Validate public scenario selectors shared by simulations and comparisons."""
    mode = request.get("mode", "us-only")
    objective = request.get("foreignObjective", "prosperity")
    pause_unavailable = request.get("pauseUnavailable", False)
    if mode not in ("us-only", "strategic"):
        raise ValueError("The scenario mode must be us-only or strategic.")
    if objective not in ("workers", "prosperity", "output"):
        raise ValueError("The foreign objective must be workers, prosperity, or output.")
    if not isinstance(pause_unavailable, bool):
        raise ValueError("pauseUnavailable must be a boolean.")
    return mode, objective, pause_unavailable


def solve_scenario(request: dict[str, Any]) -> dict[str, Any]:
    """Evaluate every permitted package; materialize only displayed trajectories.

    Old links may contain a pace field. It is deliberately ignored: deployment
    belongs on the complete ballot, not in the scenario assumptions.
    """
    mode, foreign_objective, pause_unavailable = scenario_settings(request)
    model = solve_model(
        request.get("inputs", {}),
        {
            "mode": mode,
            "objective": "workers",
            "foreignObjective": foreign_objective,
            "pace": 1,
        },
    )
    policies = [policy for policy in POLICIES if not pause_unavailable or policy["pace"] != 0]
    foreign_policies = [
        policy for policy in model["foreignPolicies"] if not pause_unavailable or policy["pace"] != 0
    ]
    election = solve_package_election(
        {
            **model,
            "policies": policies,
            "foreignPolicies": foreign_policies,
            "currentPolicy": current_policy(1),
        }
    )
    cache: dict[str, dict[str, Any]] = {}

    def materialize(policy: dict[str, Any]) -> dict[str, Any]:
        foreign = election.get("foreignPolicy")
        key = policy["id"] + "::" + (foreign["id"] if foreign else "none")
        if key not in cache:
            cache[key] = model["evaluate"](policy, foreign)
        return cache[key]

    by_id = {policy["id"]: policy for policy in policies}
    ranked = sorted(
        election["ballot"]["tallies"], key=lambda tally: (-tally["supportPercent"], tally["policyId"])
    )
    selected = materialize(election["usPolicy"])
    alternatives = [materialize(by_id[tally["policyId"]]) for tally in ranked[:8]]
    result = {
        "inputs": model["inputs"],
        "mode": model["mode"],
        "foreignObjective": model["foreignObjective"],
        "pauseUnavailable": pause_unavailable,
        "selected": selected,
        "statusQuo": materialize(current_policy(1)),
        "baseline": model["evaluate"](BASELINE_POLICY, BASELINE_POLICY if mode == "strategic" else None),
        "alternatives": alternatives,
        "ballot": election["ballot"],
        "selection": election["selection"],
        "policyCount": len(policies),
        "foreignPolicyCount": len(foreign_policies),
        "evaluations": election["evaluations"],
        "search": election["search"],
    }
    if alternatives:
        result["leading"] = alternatives[0]
    for key in ("foreignBestPolicy", "foreignBestResponseGain"):
        if key in election:
            result[key] = election[key]
    return result
